xds/src/main/java/io/grpc/xds/HttpConnectionManager.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import io.grpc.xds.Filter.NamedFilterConfig;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * HttpConnectionManager is a network filter for proxying HTTP requests.
+ */
+@AutoValue
+abstract class HttpConnectionManager {
+  // Total number of nanoseconds to keep alive an HTTP request/response stream.
+  abstract long httpMaxStreamDurationNano();
+
+  // Name of the route configuration to be used for RDS resource discovery.
+  @Nullable
+  abstract String rdsName();
+
+  // List of virtual hosts that make up the route table.
+  @Nullable
+  abstract ImmutableList<VirtualHost> virtualHosts();
+
+  // List of http filter configs. Null if HttpFilter support is not enabled.
+  @Nullable
+  abstract ImmutableList<NamedFilterConfig> httpFilterConfigs();
+
+  static HttpConnectionManager withRdsName(long httpMaxStreamDurationNano, String rdsName,
+      @Nullable List<NamedFilterConfig> httpFilterConfigs) {
+    checkNotNull(rdsName, "rdsName");
+    return HttpConnectionManager.create(
+        httpMaxStreamDurationNano, rdsName, null, httpFilterConfigs);
+  }
+
+  static HttpConnectionManager withVirtualHosts(long httpMaxStreamDurationNano,
+      List<VirtualHost> virtualHosts, @Nullable List<NamedFilterConfig> httpFilterConfigs) {
+    checkNotNull(virtualHosts, "virtualHosts");
+    return HttpConnectionManager.create(httpMaxStreamDurationNano, null, virtualHosts,
+        httpFilterConfigs);
+  }
+
+  // TODO(chengyuanzhang): server side RDS/RouteConfiguration spec has not been finalized yet,
```

Comment:
If the spec is not finalized, is this PR premature? i.e. could the code/implementation change a lot after the spec? Any pointers/timeline about the spec or its ETA?

Secondly: are we making a big assumption that "logic for how HCM should be parsed and validated should be the same between client and server side"? The ApiListener is gRPC-specific (unlike Envoy which doesn't have this client-side and server-side Listener distinction). Having started down this road why do we now expect that the HttpConnectionManager (which is one level below the Listener) processing should be unified?

Reply:
You are right. As mentioned in https://github.com/grpc/grpc-java/pull/8228#discussion_r646800116, we should just add the logic for requiring either RDS or inlined RouteConfiguration for all HttpConnectionManagers, no matter it is for the client or server side. Otherwise, the server side mostly just gets an empty HttpConnectionManager, which doesn't given much usefulness for the full support of HttpFilter (it only supports top-level HttpFilters in HttpConnectionManager but without overrides, as override HttpFilter configs are received in RDS/RouteConfiguration).

I've updated this PR to not bypass RDS or inlined RouteConfiguration validations for server side. This implies we are adding requirement for usages of HttpConnectionManager on the server side to contain either RDS or inlined RouteConfiguration.